api/pkg/webhooks/manager.go

```diff
@@ -0,0 +1,205 @@
+				client.GetUseDataFrom(),
+			)
+		}
+		p, err := client.Invoke(ctx, tmpPayload)
+		if err == nil {
+			responsePayloadLookup[client.GetName()] = p
+			if client.IsFinalResponse() {
+				finalResponse = p
+			}
+			continue
+		}
+		// if err is not nil, check if client is set to abort
+		if client.AbortOnFail() {
+			return onError(err)
+		}
+
+	}
+	for _, client := range asyncClients {
+		// NOTE: Currently, this will never return err since InvokeAsync always returns nil
+		if err := client.InvokeAsync(ctx, originalPayload); err != nil {
+			return onError(err)
+		}
+	}
+	if err := onSuccess(finalResponse); err != nil {
+		// If the callback fails, return the error
+		return err
+	}
+
+	return nil
+}
+
+func parseAndValidateConfig(
+	eventList []EventType,
+	webhookConfigMap map[EventType][]WebhookConfig,
+) (WebhookManager, error) {
+	eventToWHMap := make(map[EventType][]WebhookClient)
+	for _, eventType := range eventList {
+		if webhookConfigList, ok := webhookConfigMap[eventType]; ok {
+			var result []WebhookClient
+			for _, webhookConfig := range webhookConfigList {
+				if err := validateWebhookConfig(&webhookConfig); err != nil {
+					return nil, err
+				}
+				result = append(result, &simpleWebhookClient{
+					WebhookConfig: webhookConfig,
+				})
+			}
+			eventToWHMap[eventType] = result
+		}
+	}
+	for _, webhookClients := range eventToWHMap {
+		syncClients := make([]WebhookClient, 0)
+		for _, client := range webhookClients {
+			if !client.IsAsync() {
+				syncClients = append(syncClients, client)
+			}
+		}
+		if err := validateSyncClients(syncClients); err != nil {
+			return nil, err
+		}
+	}
+	return &SimpleWebhookManager{WebhookClients: eventToWHMap}, nil
+}
+
+// validateWebhookResponse ensures that the response from a webhook is either
+// a valid json object or empty str
+// This is only required for synchronous webhooks, where the webhook's response
+// may be used as input to the next webhook or in the callback function
+func validateWebhookResponse(content []byte) error {
+	if len(content) == 0 {
+		return nil
+	}
+	if json.Valid(content) {
+		return nil
+	}
+	return fmt.Errorf("webhook response is not a valid json object and not empty")
+}
+
+func validateSyncClients(webhookClients []WebhookClient) error {
+	// ensure that only 1 sync client has finalResponse set to true
+	isFinalResponseSet := false
+	for _, client := range webhookClients {
+		if client.IsFinalResponse() {
+			if isFinalResponseSet {
+				return fmt.Errorf("only 1 sync client can have finalResponse set to true")
+			}
+			isFinalResponseSet = true
+		}
+	}
+	if !isFinalResponseSet {
+		return fmt.Errorf("at least 1 sync client must have finalResponse set to true")
+	}
+	// Ensure that all useDataFrom fields exist
+	webhookNames := make(map[string]int)
+	for idx, client := range webhookClients {
+		if _, ok := webhookNames[client.GetName()]; ok {
+			return fmt.Errorf("duplicate webhook name")
+		}
```

Comment:
Hmm is this duplicate webhook name only a problem from sync webhooks? i.e. is it okay to have async webhooks with the same name?

Reply:
fixed